### DISCUSSION TITLE
How to extract Y+C and Y+Pb+Pr from an image as new separate images? Luminance and Chrominance

### DISCUSSION THREAD

**Discussion by tergan1** [ImageMagick/ImageMagick#3266] [Help]
Hello.
How to split a color image to Y+C and Y+Pb+Pr as new separate images?

Luminance only, Chrominance only
![label](https://i.imgur.com/Az5qAWW.png)

Luminance only, Pb only, Pr only
![label](https://i.imgur.com/CYd5DOL.png)

Original color images:
https://i.imgur.com/Dysy09y.png
https://i.imgur.com/evPkt19.png

What I need:
https://i.imgur.com/1RJaBGX.png
https://i.imgur.com/IMobgDv.png
https://i.imgur.com/co7ymDy.png
https://i.imgur.com/5j13lIw.png
https://i.imgur.com/necQZHQ.png

**Comment by snibgo** [comment]
"-colorspace YPbPr" converts channels to Y, Pb and Pr.

"-colorspace sRGB" converts channels to sRGB.

"-set colorspace sRGB" changes metadata only, declaring an image to be sRGB even if it isn't.

So:
```
%IMG7%magick Dysy09y.png -alpha off -colorspace YPbPr -separate -set colorspace sRGB +append ypbpr.png
```
![ypbpr](https://user-images.githubusercontent.com/33812368/108372137-67a9b200-71f6-11eb-955d-63df5fe5228f.png)

This makes 3 grayscale images, representing Y, Pb and Pr. It appends the images together.

Or:
```
%IMG7%magick Dysy09y.png -alpha off -colorspace YPbPr -channel 0 -evaluate set 50%% +channel -colorspace sRGB pbpronly.png
```
![pbpronly](https://user-images.githubusercontent.com/33812368/108372167-72644700-71f6-11eb-9b02-33145d4afe05.png)

This converts to YPbPr, sets the Y channel to 50% (ie middle luminance) and converts back to sRGB. So the result shows the chrominance.

**Reply by tergan1** [reply]
Hmm
1. Images do not match:
If I will combine the original Y and C images back, I will see the colored result as in the first message. 
If I will combine your pictures, then I will do not get the same result. (I tried -evaluate set 20%%, 30%%, 40%% too)
Y, C, Y+C (combined) --- (1st (left) original from first post, 2nd (right) from your script).
![](https://i.imgur.com/QxpxfiJ.png) ![](https://i.imgur.com/JFFBZVi.png)
![](https://i.imgur.com/Nax56sz.png) ![](https://i.imgur.com/3O8Yd5V.png)
![](https://i.imgur.com/AtUGN1x.png) ![](https://i.imgur.com/EGfNK5Q.png)

2. Pb and Pr are gray . I got the same result today using Gimp -> Colors -> Components -> Decompose. What is needed is not a "black and white" result, but a color, as in the first message with the house. If you will combine 3 gray images - you will get a gray image
3. The gray Y Pb Pr result is not individually splitted files. That is, ideally need 4 new separated files: Y, Pb, Pr, C

**Comment by snibgo** [comment]
> If I will combine your pictures, then I will do not get the same result.

How did you combine them? Please show your command. As my outputs are sRGB, they need to be converted to YPbPr before combining, then converted to sRGB afterwards.
```
%IMG7%magick ^
  Dysy09y.png ^
  -alpha off -colorspace YPbPr -channel 0 -separate +channel ^
  ( pbpronly.png -colorspace YPbPr -channel 1,2 -separate +channel ) ^
  -combine ^
  -set colorspace YPbPr ^
  -colorspace sRGB ^
  x.png
```
x.png looks the same as your input Dysy09y.png.
```
rem Compare the result with the original:

%IMG7%magick compare ^
  -metric RMSE ^
  Dysy09y.png x.png NULL:
```
The result is "32.9373 (0.00050259)", so the round-trip is almost perfect.

> That is, ideally need 4 new separated files: Y, Pb, Pr, C

What is C? How is it defined? I assumed it was defined by making the YPbPr image, setting Y to 50%, and converting to sRGB. If my assumption is wrong, please say how it is defined.

**Reply by tergan1** [reply]
> How did you combine them? Please show your command.

Photoshop - layer+linear light. This metod is working with 3 original images. I don't know how these images was generated by author

> x.png looks the same as your input Dysy09y.png.

Yes, but pbpronly.png is not the same like original https://i.imgur.com/IMobgDv.png

> What is C? How is it defined? I assumed it was defined by making the YPbPr image, setting Y to 50%, and converting to sRGB. If my assumption is wrong, please say how it is defined.

C is a chrominance. Defined here https://en.wikipedia.org/wiki/Chrominance (see picture in the article) , I don't know how it is calculating.
It looks like it is calculated differently in different programs and it depends on the program code

**Comment by snibgo** [comment]
> Photoshop - layer+linear light.

Sorry, I don't use Photoshop, so can't comment.

> It looks like it is calculated differently in different programs and it depends on the program code

It depends on the algorithm.
